### PR TITLE
Added RestTemplate SSL support

### LINKED ISSUE #3935: Client side TLS does not work with eureka client and rest template
**Describe the bug**

In spring-cloud **SR8**, we can enable [**client side tls**](https://github.com/spring-cloud/spring-cloud-netflix/pull/3865/files?file-filters%5B%5D=.java) by setting eureka.client.tls.keystoretype, eureka.client.tls.keystore, eureka.client.tls.enabled. But it cannot work in **2020.0.0-m5**. 

After some investigation I found Spring cloud **2020.0.0-m5** **removes jersey client** with this [change](https://github.com/spring-cloud/spring-cloud-netflix/commit/14dbd35b5290328af87e94ab60af54ae59967c82). Then DiscoveryClient will use **RestTemplate**, while the [RestTemplateTransportClientFactories](https://github.com/spring-cloud/spring-cloud-netflix/blob/v3.0.0-M5/spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestTemplateTransportClientFactories.java) does not accept sslContext.  And in SR8, DiscoveryClient will use JerseyClient, JersyClient can work with client tls perfectly.

I know I can mitigate this issue by adding jersey dependency in my project, but I am still wonder if it is a bug in RestTemplate

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
fixes #3935

### CODE REVIEW

File: spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestTemplateTransportClientFactory.java [hunk 1]
```diff
@@ -80,6 +123,22 @@ private RestTemplate restTemplate(String serviceUrl) {
 		return restTemplate;
 	}
 
+	private RestTemplate restTemplate() {
+		if (this.sslContext.isPresent()) {
+			SSLContext sslContext = this.sslContext.get();
+			// TODO: Using HttpClient here - maybe this should be configurable?
```
Comment (marcingrzejszczak) [inline spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestTemplateTransportClientFactory.java:129]: What do we do about this?

Comment (ryanjbaxter) [reply spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestTemplateTransportClientFactory.java:129]: Could it be a bean?  That way you can provide your own?

**Review by ryanjbaxter — APPROVED** [review]
LGMT

**Comment by marcingrzejszczak** [comment]
Done via https://github.com/spring-cloud/spring-cloud-netflix/commit/1e4594655500b31d0f7232c351748515876208b7 

Added an interface to provide an instance of the factory

```
/**
 * Supplier for the {@link ClientHttpRequestFactory} to be used by Eureka client.
 *
 * @author Marcin Grzejszczak
 * @since 3.0.0
 */
public interface EurekaClientHttpRequestFactorySupplier {

	/**
	 * Returns a constructed {@link ClientHttpRequestFactory}.
	 * @param sslContext SSL Context
	 * @param hostnameVerifier Hostname verifier
	 * @return {@link ClientHttpRequestFactory}
	 */
	ClientHttpRequestFactory get(SSLContext sslContext, @Nullable HostnameVerifier hostnameVerifier);

}
```

**Pull request closed without merging**